### PR TITLE
Deprecate cluster.routing.allocation.type

### PR DESCRIPTION
**Pull request by DaveCTurner**
Today users may select a shards allocator via the `cluster.routing.allocation.type` setting. In practice this is only used to select between the (default) desired-balance allocator and the legacy allocator. With this commit we mark this setting as deprecated to indicate that the desired-balance allocator will be the only choice in a future release.

### CODE REVIEW

**Comment by DaveCTurner** [comment]
This PR effectively reopens https://github.com/elastic/elasticsearch/issues/39464. The concerns raised by users in that issue, several years back, no longer seem to be relevant. In particular, the only known third-party allocator implementation was [Tempest](https://github.com/inthecloud247/tempest) which has not changed in a long time and does not support any recent ES versions anyway.

We could instead just deprecate `cluster.routing.allocation.type: balanced` and permit other values, but I think we should be opinionated that in a future version there will only be one allocator.

**Comment by DaveCTurner** [comment]
@elasticmachine update branch

File: server/src/main/java/org/elasticsearch/cluster/ClusterModule.java [hunk 1]
```diff
@@ -402,6 +404,7 @@ private static ShardsAllocator createShardsAllocator(
             });
         }
         String allocatorName = SHARDS_ALLOCATOR_TYPE_SETTING.get(settings);
+        assert Version.CURRENT.major == Version.V_7_17_0.major + 1; // in v9 there is only one allocator
```
Comment (idegtiarenko) [inline server/src/main/java/org/elasticsearch/cluster/ClusterModule.java:407]: I think we need to keep this property for as long we have a plugin API to supply a custom allocator implementation

https://github.com/elastic/elasticsearch/blob/8b92cd4bf59b2e7e99647308ad3d13cd247dbbf0/server/src/main/java/org/elasticsearch/cluster/ClusterModule.java#L385-L405

Comment (idegtiarenko) [reply server/src/main/java/org/elasticsearch/cluster/ClusterModule.java:407]: Based on above I would suggest to emit the waring only when "balanced" is used.

**Comment by DaveCTurner** [comment]
We discussed in another channel and agreed to deprecate the whole setting.

**Review by idegtiarenko — APPROVED** [review]

**Comment by DaveCTurner** [comment]
Thanks Ievgen!

**Comment by tinder-igorsokolov** [comment]
hi everyone, I'm wondering if this change means that you do not want to allow providing custom shard allocation strategy? We have a number of cluster that very read heavy and fluctuating quite well through the day. So we came up with an external shard allocator. I was considering to move that logic into ES but then I came across the deprecation annotation in the code and found out this PR. Does it mean I won't be able to move that logic inside Elasticsearch?

**Comment by DaveCTurner** [comment]
That's correct @tinder-igorsokolov, we expect that there will ultimately be just one allocator. A "fluctuating" cluster suggests a misconfiguration elsewhere in the system, and it's unlikely that adjusting or replacing the allocator is the right approach to fixing it. If you need help troubleshooting things further then please head over to the [community forums](https://discuss.elastic.co/c/elastic-stack/elasticsearch/6) and describe your problem more fully.

**Pull request merged**